### PR TITLE
Copy set before iterating, and remove guarded set

### PR DESCRIPTION
**Pull request by methane**

### CODE REVIEW

File: Modules/_abc.c [hunk 1]
```diff
@@ -838,6 +693,62 @@ _abc_subclasscheck(PyObject *m, PyObject *args)
...
+    Py_ssize_t i = 0;
+
+    while (_PySet_NextEntry(impl->_abc_registry, &pos, &key, &hash)) {
+        Py_INCREF(key);
+        copy[i++] = key;
+    }
+    assert(i == registry_size);
+
+    for (i = 0; i < registry_size; i++) {
+        PyObject *rkey = PyWeakref_GetObject(copy[i]);
```
Comment (pppery) [inline Modules/_abc.c:731]: copy[i] isn't necessary a weakref; if someone uses `weakref.getweakrefs` to access the internal registry, as I explained on the main PR. In this case, rkey would then be null.

**Comment by ilevkivskyi** [comment]
Thank you! It looks like this will be indeed safe (even in multithreaded context). However, I would like to consider other consequences.
Positive:
* This simplifies the overall logic.
* This simplifies implementation.

Negative:
* It looks like this can have impact on performance. Note that this code will copy registry on every `issubclass()` and `isinstance()` call (also internally). Of course we can rely on caches for this, but I am still not sure.

Also maybe someone knows why such a complex solution (guarded weak set) was chosen in the original pure Python implementation? My suspicion is it may be for performance reasons.

**Comment by ilevkivskyi** [comment]
cc @1st1
> Thank you! It looks like this will be indeed safe (even in multithreaded context). However, I would like to consider other consequences.
Positive:
*This simplifies the overall logic.
*This simplifies implementation.

> Negative:
*It looks like this can have impact on performance. Note that this code will copy registry on every issubclass() and isinstance() call (also internally). Of course we can rely on caches for this, but I am still not sure.

> Also maybe someone knows why such a complex solution (guarded weak set) was chosen in the original pure Python implementation? My suspicion is it may be for performance reasons.

**Comment by pppery** [comment]
... because `weakref.WeakSet` already existed in the standard library?

File: Modules/_abc.c (hunk 1 above)
Comment (pppery) [reply Modules/_abc.c:731]: causing a crash in the Py_INCREF on the next line

Comment (methane) [reply Modules/_abc.c:731]: CPython is not sand-boxed language like JavaScript.
Basic operation and metaprogramming shouldn't cause crash.
But we don't have to guard from people break CPython via gc, ctypes, weakref, etc...

But in this case, adding `rkey == NULL` check is easy and natural.
I added it, thank you.

**Comment by methane** [comment]
Iterating set and copy are both O(n).  So I don't think this copy can make ABC
slower than Python version even if there are many types in the registry.

And I added fast path too.

**Comment by ilevkivskyi** [comment]
> So I don't think this copy can make ABC
slower than Python version even if there are many types in the registry.

Yes, sure, I meant it may be _a bit_ slower than the C implementation with guards. Since the performance implications are minor, I am in favour of using this. So unless @1st1 has some concerns, I will use this instead of iteration guards.

**Comment by methane** [comment]
Ping @1st1

**Comment by methane** [comment]
I want one more idea: Create set just in time.  registry is created when subclass is registered.  negative cache is created when subclasscheck is called.

Empty set is 224byte on 64bit Linux.  So it will reduce 224*3=672bytes for each class.
This is important because all classes inheriting ABC is ABC, even if they are concrete class.

**Comment by ilevkivskyi** [comment]
> Create set just in time. registry is created when subclass is registered. negative cache is created when subclasscheck is called.

Yes, but this is independent and can be done _after_ this lands.

**Comment by ilevkivskyi** [comment]
> Ping @1st1

It looks like @1st1 is busy now. Anyway, I think he is not strongly against this (otherwise he would probably said this) we should go forward with this, so I am going to merge this.

Then we can finish some details in the main PR (more tests, maybe add GC head, lazy loading of Python `ABCMeta` implementation, this will speed-up Python startup by around 1 ms).

**Comment by methane** [comment]
> maybe add GC head

I don't think `_abc_data` needs GC head.  It contains only weak references, unless people expose internal set via gc or weakref.

**Pull request merged**